kdf.js

```diff
@@ -0,0 +1,114 @@
+// const { randomBytes } = require('./random-bytes')
+const isBuffer = require('is-buffer')
+
+/* eslint-disable camelcase */
+const {
+  crypto_kdf_CONTEXTBYTES,
+  crypto_kdf_BYTES_MAX,
+  crypto_kdf_BYTES_MIN,
+  crypto_kdf_PRIMITIVE,
+  crypto_kdf_KEYBYTES,
+
+  crypto_kdf_derive_from_key,
+  crypto_kdf_keygen,
+} = require('./sodium')
+
+/**
+ * Generates a master key.
+ *
+ * @public
+ * @param {?(Buffer)} [key]
+ * @returns {Buffer}
+ * @throws TypeError
+ */
+function keygen(key) {
+  if (key && false === isBuffer(key)) {
+    throw new TypeError('kdf.keygen: Expecting key to be a buffer.')
+  }
+
+  if (undefined === key) {
+    key = Buffer.allocUnsafe(crypto_kdf_KEYBYTES)
+  }
+
+  if (crypto_kdf_KEYBYTES !== key.length) {
+    throw new TypeError(`kdf.keygen: Invalid key length: ${key.length}`)
+  }
+
+  crypto_kdf_keygen(key)
+  return key
+}
+
+/**
+ * Derives a subkey using the master key and context.
+ *
+ * @public
+ * @param {?(Buffer)} [subkey]
+ * @param {?(Number)} [subkeyId]
+ * @param {Buffer} ctx
+ * @param {Buffer} key
+ * @return {Buffer}
+ * @throws TypeError
+ */
+function derive(subkey, subkeyId, ctx, key) {
```

Comment:
Perhaps this function should be a bit lower level and a higher level one could exist that simply allows one to give a `key` from `kdf.keygen()` and pass it to a `kdf.derive(key, iterations)` that returns the final subkey. I imagine the `derive` function you currently have being broken up into a `ctx = kdf.init(key)`, `subkey kdf.update(ctx, id)`, and `kdf.final(ctx)` which `kdf.derive()` should compose. A simple an unchecked example can be seen below:

```js
function init(key) {
  return { key, buffer: Buffer.allocUnsafe(crypto_kdf_CONTEXTBYTES), subkey: null }
}

function update(ctx, id) {
  ctx.subkey = ctx.subkey || Buffer.allocUnsafe(crypto_kdf_KEYBYTES)
  crypto_kdf_derive_from_key(ctx.subkey, id, ctx.buffer, ctx.key)
  return ctx.subkey
}

function final(ctx) {
  const subkey = ctx.subkey
  ctx.subkey = null
  return ctx.subkey
}

function derive(key, iterations) {
  const ctx = init(key)
  for (let i = 0; i < iterations.length; ++i) {
    update(ctx, i) // or i + i
  }
  return final(ctx)
}
```

Reply:
Ok, I'll break it down into smaller parts.

Reply:
Well, `init()` could be rewritten to account for a given context buffer and already initialized `subkey`:

```js
function init(key, buffer, subkey)
  return {
    key,
    buffer: buffer || Buffer.allocUnsafe(crypto_kdf_CONTEXTBYTES),
    subkey: subkey || null }
}
```

Reply:
@jwerle Do you want me to disregard stylistically how all the other files do error handling or would you prefer I stay consistent while incorporating your suggestions?

Reply:
The latter please. Consistent is better 🙃

Reply:
👍